providers/gce/gce_loadbalancer.go

```diff
@@ -197,19 +210,76 @@ func (g *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, svc
 		}
 	}
 
-	var status *v1.LoadBalancerStatus
+	var syncResult *LBSyncResult
 	switch desiredScheme {
 	case cloud.SchemeInternal:
-		status, err = g.ensureInternalLoadBalancer(clusterName, clusterID, svc, existingFwdRule, nodes)
+		syncResult, err = g.ensureInternalLoadBalancer(clusterName, clusterID, svc, existingFwdRule, nodes)
 	default:
-		status, err = g.ensureExternalLoadBalancer(clusterName, clusterID, svc, existingFwdRule, nodes)
+		syncResult, err = g.ensureExternalLoadBalancer(clusterName, clusterID, svc, existingFwdRule, nodes)
 	}
 	if err != nil {
 		klog.Errorf("Failed to EnsureLoadBalancer(%s, %s, %s, %s, %s), err: %v", clusterName, svc.Namespace, svc.Name, loadBalancerName, g.region, err)
-		return status, err
+		return syncResult.Status, err
+	}
+
+	if g.enableL4LBAnnotations {
+		if err = g.updateL4ResourcesAnnotations(ctx, svc, syncResult.Annotations); err != nil {
+			return syncResult.Status, fmt.Errorf("failed to set resource annotations, err: %w", err)
+		}
 	}
+
 	klog.V(4).Infof("EnsureLoadBalancer(%s, %s, %s, %s, %s): done ensuring loadbalancer.", clusterName, svc.Namespace, svc.Name, loadBalancerName, g.region)
-	return status, err
+	return syncResult.Status, err
+}
+
+func (g *Cloud) updateL4ResourcesAnnotations(ctx context.Context, svc *v1.Service, newL4LBAnnotations map[string]string) error {
+	newObjectMeta := computeNewAnnotationsIfNeeded(svc, newL4LBAnnotations, L4ResourceAnnotationKeys)
+	if newObjectMeta == nil {
+		return nil
+	}
+	return g.PatchServiceObjectMetadata(ctx, svc, *newObjectMeta)
+}
+
+// PatchServiceObjectMetadata patches the given service's metadata based on new
+// service metadata.
+func (g *Cloud) PatchServiceObjectMetadata(ctx context.Context, svc *v1.Service, newObjectMetadata metav1.ObjectMeta) error {
+	newSvc := svc.DeepCopy()
+	newSvc.ObjectMeta = newObjectMetadata
+	_, err := g.PatchService(ctx, svc, newSvc)
+	return err
+}
+
+// PatchService patches the given service's Status or ObjectMeta based on the original and
+// updated ones. Change to spec will be ignored.
+func (g *Cloud) PatchService(ctx context.Context, oldSvc, newSvc *v1.Service) (*v1.Service, error) {
+	// Reset spec to make sure only patch for Status or ObjectMeta.
+	newSvc.Spec = oldSvc.Spec
+
+	patchBytes, err := getPatchBytes(oldSvc, newSvc)
+	if err != nil {
+		return nil, err
+	}
+
+	return g.client.CoreV1().Services(oldSvc.Namespace).Patch(context.TODO(), oldSvc.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
```

Comment:
Let's use the context that is passed here instead of a new one.

```suggestion
	return g.client.CoreV1().Services(oldSvc.Namespace).Patch(ctx, oldSvc.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
```